### PR TITLE
Adaptive and Fed DG-GA PFL Experimentation

### PR DESCRIPTION
**Pull request by emersodb**
# PR Type
Experimentation

# Short Description

Clickup Ticket(s): [Link](https://app.clickup.com/t/8689g93je), [Link](https://app.clickup.com/t/8688d097n)

This PR implements a set of scripts for running experiments on the Heterogenous CIFAR-10 datasets using the Adaptive pFL and Fed DG-GA pFL classes. These will facilitate an initial set of PoC experiments to be expanded to other datasets in followups.

# Tests Added

N/A

### CODE REVIEW

File: research/cifar10/adaptive_pfl/ditto/server.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    )
+    parser.add_argument(
+        "--seed",
+        action="store",
+        type=int,
+        help="Seed for the random number generators across python, torch, and numpy",
+        required=False,
+    )
+    parser.add_argument(
+        "--lam", action="store", type=float, help="Ditto loss weight for local model training", default=0.01
```
Comment (lotif) [inline research/cifar10/adaptive_pfl/ditto/server.py:104]: Can we spell out `lambda` instead? Just for more clarity. Here and also on the variables and parameters.

File: research/cifar10/adaptive_pfl/fedprox/client.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    parser.add_argument(
+        "--beta",
+        action="store",
+        type=float,
+        help="Heterogeneity level for the dataset",
+        required=True,
+    )
+    args = parser.parse_args()
+
+    DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (lotif) [inline research/cifar10/adaptive_pfl/fedprox/client.py:130]: Upper case var names should be reserved for global vars and constants.

File: research/cifar10/adaptive_pfl/fenda_ditto/client.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    )
+    parser.add_argument(
+        "--freeze_global_extractor",
+        action="store_true",
+        help="Whether or not to freeze the global feature extractor of the FENDA model or not.",
+        default=False,
+    )
+    args = parser.parse_args()
+
+    DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (lotif) [inline research/cifar10/adaptive_pfl/fenda_ditto/client.py:138]: Same here.

File: research/cifar10/model.py [hunk 4]
```diff
@@ -45,3 +48,128 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        h: int = 32,
+        w: int = 32,
+        hidden: int = 2048,
+        class_num: int = 10,
+        dropout: float = 0.0,
+    ) -> None:
+        super().__init__(join_mode)
+
+        # Times 2 because we'll be concatenating the inputs of two feature extractors
+        self.fc1 = Linear((h // 2 // 2) * (w // 2 // 2) * 64 * 2, hidden)
```
Comment (lotif) [inline research/cifar10/model.py:122]: Nice! I didn't know about the `//` operator.

File: research/cifar10/preprocess.py [hunk 5]
```diff
@@ -100,20 +100,20 @@ def preprocess_data(
 
     # Partition train data
     heterogeneous_partitioner = DirichletLabelBasedAllocation(
-        number_of_partitions=num_clients, unique_labels=list(range(10)), beta=beta, min_label_examples=2
+        number_of_partitions=num_clients, unique_labels=list(range(10)), beta=beta, min_label_examples=1
     )
     train_partitioned_datasets, train_partitioned_dist = heterogeneous_partitioner.partition_dataset(
-        training_set, max_retries=5
+        training_set, max_retries=-1
```
Comment (lotif) [inline research/cifar10/preprocess.py:106]: What does `-1` retries mean?

File: research/cifar10/adaptive_pfl/ditto/server.py (hunk 1 above)
Comment (emersodb) [reply research/cifar10/adaptive_pfl/ditto/server.py:104]: Will do.

File: research/cifar10/adaptive_pfl/fedprox/client.py (hunk 2 above)
Comment (emersodb) [reply research/cifar10/adaptive_pfl/fedprox/client.py:130]: Yeah that's a good point. I think we do this in a bunch of places in the library and should probably drop the practice. I'll open up another PR that drops these to lowercase everywhere.

File: research/cifar10/adaptive_pfl/fenda_ditto/client.py (hunk 3 above)
Comment (emersodb) [reply research/cifar10/adaptive_pfl/fenda_ditto/client.py:138]: 👍  See other comment.

File: research/cifar10/model.py (hunk 4 above)
Comment (emersodb) [reply research/cifar10/model.py:122]: Credit to @sanaAyrml for this usage.

File: research/cifar10/preprocess.py (hunk 5 above)
Comment (emersodb) [reply research/cifar10/preprocess.py:106]: I'll put a note in the documentation. It effectively means infinite. I'm kind of "abusing" how the underlying code works to get this functionality.

File: research/cifar10/adaptive_pfl/fedprox/client.py (hunk 2 above)
Comment (emersodb) [reply research/cifar10/adaptive_pfl/fedprox/client.py:130]: New PR making this change across the library is here: https://github.com/VectorInstitute/FL4Health/pull/279

File: research/cifar10/preprocess.py (hunk 5 above)
Comment (lotif) [reply research/cifar10/preprocess.py]: I think it's fine to use it like that, having it in the documentation helps understanding the use so I think we should do it.

**Review by lotif — APPROVED** [review]

**Pull request merged**